### ISSUE TITLE
Poisson function failed on rate vector contain zeros. 

### ISSUE THREAD

**Issue by Hananel-Hazan** [pytorch/pytorch#53485]
## 🐛 Bug

Poisson function failed on rate vector contain zeros. 

## To Reproduce

Run the following code on 
```
r = torch.tensor([0, 0.1, 0.2, 0.3])
p = torch.distributions.poisson.Poisson(rate=r)
```
PyTorch 1.7.1  run fine
PyTorch 1.8  getting ``Exception has occurred: ValueError
The parameter rate has invalid values``

## Expected behavior

Zeros should be acceptable value

## Environment
```
PyTorch version: 1.8.0+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3

Python version: 3.8 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.20.1
[pip3] torch==1.8.0+cpu
[pip3] torchaudio==0.8.0
[pip3] torchvision==0.9.0+cpu
[conda] Could not collect
```
OR 
```
PyTorch version: 1.7.1
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: Could not collect

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce RTX 2070
Nvidia driver version: 460.39
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.4
[pip3] torch==1.7.1
[pip3] torchvision==0.8.2
[conda] Could not collect
```

cc @fritzo @neerajprad @alicanb @vishwakftw @nikitaved

**Comment by gchanan** [comment]
I could reproduce this; it's unclear to me if this is intentional or not.

**Comment by gchanan** [comment]
The source is https://github.com/pytorch/pytorch/pull/48743 and it looks intentional -- the documentation shows:
```
arg_constraints = {'RATE': GREATERTHAN(LOWER_BOUND=0.0)}
```

**Comment by fritzo** [comment]
Hmm @neerajprad @fehiepsi I guess we could relax this to `arg_constraints = {'rate': constraints.nonnegative}`, though gradients through `.log_prob()` may NAN.

@Hananel-Hazan FYI you can work around this by disabling validation on this one distribution `Poisson(rate, validate_args=False)`

**Comment by fehiepsi** [comment]
I prefer using `validate_args=False` because Poisson with `rate=0` is not a proper distribution.

**Comment by Hananel-Hazan** [comment]
Thanks @fritzo for the work around. `arg_constraints = {'rate': constraints.nonnegative}` as you suggested is more backward compatibility and supported by the documentation ` arg_constraints = {'rate': GreaterThan(lower_bound=0.0)}` and also in the [example](https://pytorch.org/docs/stable/generated/torch.poisson.html)

**Comment by fritzo** [comment]
@Hananel-Hazan can you explain which methods of the `Poisson(0)` distribution you are using?  I ask because PyTorch does not (and never has) supported computing `Poisson(0).log_prob(x)`, but it looks like `Poisson(0,validate_args=False).sample()` does work.  This makes me inclined to agree with @fehiepsi that we keep the check as is to prevent NANs from being emitted by `.log_prob()`.  Note if you simply want a sampler you can call `torch.poisson()`.

**Comment by Hananel-Hazan** [comment]
@fritzo I'm using `torch.distributions.poisson.Poisson` to generates Poisson-distributed spike trains based on input intensity [link](https://github.com/BindsNET/bindsnet/blob/2eb3aa5d2bb187c586cff45daed0a244ca6003d0/bindsnet/encoding/encodings.py#L110).

Turning the `validate_args` on by default #48743 is the right move. The error need to come from using `.log_prob` with zero input, because zero is not a valid input for the log function by definition. But Poisson with `rate=0` does not violating the function definition and should return zero, however negative number does violent the definition. Using `validate_args=False` can cause to an opening for errors without justification. Relaxing the `'rate': constraints.nonnegative` as you suggested is justify here.

**Comment by tillahoffmann** [comment]
It would be nice to have `rate = 0` supported. While simply plugging in a zero rate into the common `log_prob` expression yields a `nan`, the distribution is properly defined in the limit `rate -> 0`: it is a point mass at zero. The expected behaviour would be

```python
>>> d = torch.distributions.Poisson(0.0)
>>> d.log_prob(torch.tensor(0))
0
>>> d.log_prob(torch.tensor(1))
-inf
```
